### ISSUE TITLE
route_prefix doesn't allow the prefix pattern to match without a trailing slash

### ISSUE THREAD

**Issue by mmerickel** [Pylons/pyramid#406]
The issue here is an ambiguity in handling `route_prefix`. In general, things work correctly:

``` python
def routes(config):
    config.add_route('add_user', '/add')

config.include(routes, route_prefix='/users')

# resulting pattern: '/users/add'
```

The ambiguity arises when the included function attempts to add a route using the pattern `'/'` or `''` (an empty string). Pyramid elects to treat the prefix as a container.

``` python
def routes(config):
    config.add_route('users', '/')

config.include(routes, route_prefix='/users')

# resulting pattern: '/users/'
```

Remember that in Pyramid's `config.add_route` there is no difference between prepending a `'/'` or leaving no prefix. Thus, the two route definitions below result in the same url being matched and generated.

``` python
config.add_route('r1', '/route')
config.add_route('r2', 'route')
```

There are 4 possible use cases to account for when looking at the issue.

**Case 1**

``` python
def routes(config):
    config.add_route('route', '')

config.include(routes, route_prefix='/prefix')
```

**Case 2**

``` python
def routes(config):
    config.add_route('route', '/')

config.include(routes, route_prefix='/prefix')
```

**Case 3**

``` python
def routes(config):
    config.add_route('route', '')

config.include(routes, route_prefix='/prefix/')
```

**Case 4**

``` python
def routes(config):
    config.add_route('route', '/')

config.include(routes, route_prefix='/prefix/')
```

**Workaround**

Pyramid currently treats the route prefix as a container, thus the resulting route will always be appended with a `'/'`. This may not be ideal, but it at least allows for a workaround where the user can add their own route (at the level of the include):

``` python
def routes(config):
    config.add_route('add', '/add')

config.include(routes, route_prefix='/users')
config.add_route('users', '/users')
```

If there is a clear way to handle each use-case, and document it as such, then it may be possible to change this behavior but to me it's hard to expect a user to "do the right thing" with case 2 due to the way pyramid currently handles `add_route` patterns with and without a prefixed slash.

**Comment by sontek** [comment]
@mmerickel When you going to fix this so that #406 will just work? =)

**Comment by bertjwregeer** [comment]
I have a feeling that this new bug report is related: https://github.com/Pylons/pyramid/issues/2143

**Comment by jvanasco** [comment]
Regarding the 2.0 (#2362) possibility:

I think @mmerickel's original notion of making things dependent on the existence of a trailing slash in the `config.include` is the correct option, but that discussion is 4 years old and so even more backwards compatible now. 

I think think the simplest option is to addd a kwarg to `config_include` that could disable the trailing slash on indexes (but defaults to keeping it for backwards compatibility).

* The `AppendSlashNotFoundFactory` is not a good solution, because it applies globally - not just on a particular route/component.
* The `pyramid_rewrite` is an external library/not-core and is not guaranteed to work across versions.

Keep in mind that different plugins will be built and want to be implemented differently.  One plugin might not even have a '' route, others might.  Some people might not even be using the `route_prefix` functionality to build plugins, but just to customize how their app mounts a particular section.

The best workaround I've found in the meantime is to explicitly declare whatever the "index" route is named to be just the prefix:

    def includeme(config):
        route_prefix = config.registry.settings.get("admin_prefix")
        config.include(_admin_views, route_prefix=route_prefix)
        config.add_route('admin-index', route_prefix)

    def _admin_views(config):
        config.add_route('admin-index', '')

That requires no 3rd party module, and does not introduce a global not-found behavior.

An implementation detail of routes that I noticed (but haven't had time to look into), is that this method will always result in the `config.add_route('admin', route_prefix)` to be the working route regardless of order (either before or after the include).
